### ISSUE TITLE
[docker-worker] race between worker shutdown and claimWork calls

### ISSUE THREAD

**Issue by djmitche** [taskcluster/taskcluster#3080]
https://community-tc.services.mozilla.com/tasks/IpXNPp4ISw6kJTpj4vO_zg/runs/0

Reading the logs:

```
Jun 16 23:23:46 proj-deepspeech-ci-jo4e5iymsnmtyfsksydsqg docker-worker: 2020/06/16 23:23:38 {"type":"task resolved","source":"top","provisionerId":"proj-deepspeech","workerId":"6851375838966116050","workerGroup":"us-east1","workerType":"ci","workerNodeType":"projects/757942385826/machineTypes/n1-standard-32","taskId":"QqlcZu3rTCWuix7L2ec4Eg","runId":1,"taskState":"completed"}
Jun 16 23:23:46 proj-deepspeech-ci-jo4e5iymsnmtyfsksydsqg docker-worker: 2020/06/16 23:23:38 WORKER_METRICS {"eventType":"taskFinish","worker":"docker-worker","workerPoolId":"proj-deepspeech/ci","workerId":"6851375838966116050","timestamp":1592349818,"region":"us-east1","instanceType":"projects/757942385826/machineTypes/n1-standard-32","taskId":"QqlcZu3rTCWuix7L2ec4Eg","runId":1} 

Jun 16 23:23:46 proj-deepspeech-ci-jo4e5iymsnmtyfsksydsqg docker-worker: 2020/06/16 23:23:38 {"EnvVersion":"2.0","Fields":{"key":"capacity.runningTasks","v":1,"val":1},"Hostname":"proj-deepspeech-ci-jo4e5iymsnmtyfsksydsqg","Logger":"taskcluster.docker-worker.proj-deepspeech.ci.projects/757942385826/machineTypes/n1-standard-32","Pid":3508,"Severity":6,"Timestamp":1592349818082000000,"Type":"monitor.count","serviceContext":{"service":"docker-worker","version":"30.1.1"},"severity":"INFO"} 
```
so QqlcZu3rTCWuix7L2ec4Eg (the previous task for this worker) is resolved at 23:23:46.

```
Jun 16 23:24:02 proj-deepspeech-ci-jo4e5iymsnmtyfsksydsqg docker-worker: 2020/06/16 23:24:02 {"type":"pending shutdown","source":"top","provisionerId":"proj-deepspeech","workerId":"6851375838966116050","workerGroup":"us-east1","workerType":"ci","workerNodeType":"projects/757942385826/machineTypes/n1-standard-32","time":15} 
```
16s later [`onIdle` logs "pending shutdown"](https://github.com/taskcluster/taskcluster/blob/138611a01fe039ff0ff3b591c6f31d6d70e29ad1/workers/docker-worker/src/lib/shutdown_manager.js#L42-L57) and starts a 15s timer.  The worker was truly idle at this point -- the next task wasn't scheduled until 2020-06-16T23:24:18.261Z

```
Jun 16 23:24:17 proj-deepspeech-ci-jo4e5iymsnmtyfsksydsqg docker-worker: 2020/06/16 23:24:17 {"EnvVersion":"2.0","Fields":{"key":"shutdown","v":1,"val":1},"Hostname":"proj-deepspeech-ci-jo4e5iymsnmtyfsksydsqg","Logger":"taskcluster.docker-worker.proj-deepspeech.ci.projects/757942385826/machineTypes/n1-standard-32","Pid":3508,"Severity":6,"Timestamp":1592349857059000000,"Type":"monitor.count","serviceContext":{"service":"docker-worker","version":"30.1.1"},"severity":"INFO"}
Jun 16 23:24:17 proj-deepspeech-ci-jo4e5iymsnmtyfsksydsqg docker-worker: 2020/06/16 23:24:17 {"type":"shutdown","source":"top","provisionerId":"proj-deepspeech","workerId":"6851375838966116050","workerGroup":"us-east1","workerType":"ci","workerNodeType":"projects/757942385826/machineTypes/n1-standard-32"}
Jun 16 23:24:17 proj-deepspeech-ci-jo4e5iymsnmtyfsksydsqg docker-worker: 2020/06/16 23:24:17
Jun 16 23:24:17 proj-deepspeech-ci-jo4e5iymsnmtyfsksydsqg docker-worker: 2020/06/16 23:24:17 WORKER_METRICS {"eventType":"instanceShutdown","worker":"docker-worker","workerPoolId":"proj-deepspeech/ci","workerId":"6851375838966116050","timestamp":1592349857,"region":"us-east1","instanceType":"projects/757942385826/machineTypes/n1-standard-32"}
Jun 16 23:24:17 proj-deepspeech-ci-jo4e5iymsnmtyfsksydsqg docker-worker: 2020/06/16 23:24:17 {"type":"exit","source":"top","provisionerId":"proj-deepspeech","workerId":"6851375838966116050","workerGroup":"us-east1","workerType":"ci","workerNodeType":"projects/757942385826/machineTypes/n1-standard-32"} 
```
15 seconds later (and one second before IpXNPp4ISw6kJTpj4vO_zg becomes pending), we begin to shut down.  There are no intervening logs about starting a task.  This triggers a call to `workerManager.removeWorker` by worker-runner, which eventually powers-down the machine:
```
Jun 16 23:24:18 proj-deepspeech-ci-jo4e5iymsnmtyfsksydsqg systemd-logind: Power key pressed. 
```
a few seconds later, we successfully claim IpXNPp4ISw6kJTpj4vO_zg:
```

Jun 16 23:24:22 proj-deepspeech-ci-jo4e5iymsnmtyfsksydsqg docker-worker: 2020/06/16 23:24:19 {"type":"claimed task","source":"top","provisionerId":"proj-deepspeech","workerId":"6851375838966116050","workerGroup":"us-east1","workerType":"ci","workerNodeType":"projects/757942385826/machineTypes/n1-standard-32","taskId":"IpXNPp4ISw6kJTpj4vO_zg","runId":0}
Jun 16 23:24:22 proj-deepspeech-ci-jo4e5iymsnmtyfsksydsqg docker-worker: 2020/06/16 23:24:19 {"type":"not superseding","source":"top","provisionerId":"proj-deepspeech","workerId":"6851375838966116050","workerGroup":"us-east1","workerType":"ci","workerNodeType":"projects/757942385826/machineTypes/n1-standard-32","taskId":"IpXNPp4ISw6kJTpj4vO_zg","message":"no supersederUrl in payload"}
Jun 16 23:24:22 proj-deepspeech-ci-jo4e5iymsnmtyfsksydsqg docker-worker: 2020/06/16 23:24:19 {"type":"run task","source":"top","provisionerId":"proj-deepspeech","workerId":"6851375838966116050","workerGroup":"us-east1","workerType":"ci","workerNodeType":"projects/757942385826/machineTypes/n1-standard-32","taskId":"IpXNPp4ISw6kJTpj4vO_zg","runId":0} 
```
and *that* is the task that is cut off with worker-terminated.

So I think this is a race condition.  I think what's happened here is that docker-worker has called `queue.claimWork`, which long-polls for 20s before returning 0 results.  And docker-worker waits for 5s between calls to `claimWork`.  So at 23:24:02, claimWork returned 0 tasks.  5s later, at 23:24:07, docker-worker would have called `claimWork` again (it doesn't log this, and queue logs the *response* time of an HTTP call, so this is conjecture).  Then less than 20s later at 23:24:22 that `claimWork` returned a task.  The task listener had been closed, but it was already in the middle of the claimWork call at the time.

Fundamentally, we really can't get rid of this race without some kind of two-phase commit: even if docker-worker were to abort the claimWork HTTP transaction, the queue might have already marked the task as claimed and sent it to the worker.  But that race is on the scale of milliseconds.  We can certainly address the 20s race, by only actually shuttting down when there's not a claimWork call in flight.  The code is brittle so that will be a difficult change to make, though.

**Comment by djmitche** [comment]
It seems that this race condition would be tickled by idle gaps that happen to last about 10s or 30s.  I'm not sure if the pattern of deepspeech jobs fits that description more often than other projects.

One workaround would be to increase the idle timeout from 15s to something a bit larger.  That only helps in the sense that it makes it less likely that a task will arrive during that 25 window of vulnerability.

**Comment by djmitche** [comment]
https://papertrailapp.com/systems/ip-172-31-42-217/events?q=program%3Adocker-worker was configured with `afterIdleSeconds: 60` and successfully shut down after that time.

It's worth noting for those counting seconds that this fix results in an average of 12.5 seconds additional idle time, as it means we only shut down at a particular point in the 25-second work-claiming loop.
